### ISSUE TITLE
pyright doesn't understand int/bool overlap

### ISSUE THREAD

**Issue by GBeauregard** [microsoft/pyright#3228]
**Describe the bug**
Pyright wrongly evaluates expressions with int/bool overlap

**To Reproduce**
```python
if True == 1:  # "Condition will always evaluate to False"
    print("hi")
```

**Expected behavior**
The condition will always evaluate to True, but pyright states it will always evaluate False.

**VS Code extension or command-line**
pyright 1.1.231 commandline

**Additional context**
This is a contributing factor to #3214

**Comment by erictraut** [comment]
This would require some specific special-casing logic in pyright. I don't know if it's worth it to handle a case that is 1) arguably bad code (shouldn't be comparing bool and int values in the first place), and 2) really easy to work around in the the code (by making the values of comparable type).

**Comment by JelleZijlstra** [comment]
Does your overlapping logic look at the way the `__eq__` method is annotated?

Perhaps it could be implemented as giving an error for `L == R` if `R` does not overlap with the argument to `L.__eq__` and `R` does not overlap with the argument to `L.__eq__`. (But perhaps with a special case if L or R doesn't override `object.__eq__`.) Then we could fix this issue by annotating `bool.__eq__` in typeshed as `def __eq__(self, obj: int) -> bool: ...`.

**Comment by erictraut** [comment]
No, the logic doesn't take into account `__eq__` because `__eq__` invariably accepts `object`, so it can't be used as a descriminator.

**Comment by erictraut** [comment]
This diagnostic rule is off by default and is enabled only in "strict" mode. The philosophy for strict-mode rules is that your code needs to be written to accommodate the rule, not the other way around. So I don't think it makes sense to add this special case logic to accommodate what I consider to be a bad coding practice.

**Comment by JelleZijlstra** [comment]
> No, the logic doesn't take into account `__eq__` because `__eq__` invariably accepts `object`, so it can't be used as a descriminator.

We can change that though, if we decide we can put some useful meaning into the annotation for `__eq__`. I'd be happy to change typeshed for that.

How does pyright know at the moment that something like `int() == float()` is fine? If it's through special casing, then my suggestion above would allow you to remove the special casing.
